README.md
# HALO reader – tool for HALO Photonics wind doppler lidar post-processing

[![Tests](https://github.com/actris-cloudnet/halo-reader/actions/workflows/ci.yml/badge.svg)](https://github.com/actris-cloudnet/halo-reader/actions/workflows/ci.yml)
[![PyPI version](https://badge.fury.io/py/halo-reader.svg)](https://badge.fury.io/py/halo-reader)

**Package is under development and functionality/features might change.**

**Feedback is welcome:**
[mail](mailto:actris-cloudnet@fmi.fi) |
[create issue](https://github.com/actris-cloudnet/halo-reader/issues/new)


## Features

* Read and merge raw files into a netCDF file
* Read raw data directly from [Cloudnet](https://cloudnet.fmi.fi/) into a netCDF
* Visualise data


## Installation

```bash
# Requires Python 3.10 or newer

# create a virtual environment (optional)
python3 -m venv env
# activate the virtual environment
source env/bin/activate
# update pip
pip install -U pip

# install
pip install halo-reader
```

### Notes for Windows Subsystem for Linux (WSL) users

You can get Python 3.10 by installing [Ubuntu 22.04 from microsoft store](https://apps.microsoft.com/store/detail/ubuntu-22042-lts/9PN20MSR04DW).
In WSL, you may need to install `build-essential` and `python3-dev` before installing `halo-reader`.


## Usage

```bash
haloreader --help
haloreader from_raw --help
haloreader from_cloudnet --help
```

### Use data from cloudnet

```bash
# generate halo_warsaw_2023-03-16.nc file
# and vis/halo_warsaw_2023-13-16.png visualisation
haloreader from_cloudnet --site warsaw --date 2023-03-16 --plot

# Browse generated visualisations at /vis directory
haloboard
# open your browser at localhost:5000
```

### Use raw files

```bash
haloreader from_raw Stare_213_20230326_*.hpl Background_*0323-*.txt -o out.nc --plot

# Browse generated visualisations at /vis directory
haloboard
# open your browser at localhost:5000
```
**Note that a good background correction requires around 300 or more background profiles.**

## License

MIT


CHANGELOG.md
# Changelog

## [Unreleased]

### Added

### Changed

### Deprecated

### Removed

### Fixed

## [0.1.9] - 2023-11-16

### Fixed
- Fixed globbing

## [0.1.8] - 2023-10-17

### Changed
- Ignore lower gates based on range insted off simply masking first 3

### Fixed
- In halodata/cli ignore records with cross tag

## [0.1.7] - 2023-10-12

### Added
- HaloException

## [0.1.6] - 2023-10-11

### Changed
- Cython reader for bg files without newlines
- Use the most common number of gates when merging files
- Filter zero bg files

## [0.1.5] - 2023-09-22

### Changed
- Skip files with different number of gates

## [0.1.4] - 2023-09-15

### Fixed
- Handle background file with zeros
- Support background file without newlines

## [0.1.3] - 2023-09-14

### Fixed
- Allow missing pitch and roll

## [0.1.2] - 2023-04-28

### Added
- Raise error if range gate indexing is inconsitent
- Add plotting for beta and screened variables

### Changed
- Screen values where intensity is below one

## [0.1.1] - 2023-04-25

### Changed
- Skip invalid input files in `read` function

## [0.1.0] - 2023-04-20

### Changed
- Ignore profiles where timestamps are not increasing

### Fixed

## [0.0.9] - 2023-04-13

### Changed
- default wavelength from 1.5um to 1.565um

## [0.0.8] - 2023-04-12

### Added
- Add posibility to change variable and attribute names in netCDF
- Exclude variables from netCDF
- Add screened beta and doppler velocity

## [0.0.7] - 2023-04-03

### Added
- Add nu and eta for beta (still for testing)

## [0.0.6] - 2023-03-31

### Added
- Compute beta using placeholder values (for testing)

### Changed
- Use float instead of douple in data arrays (excluding time)

## [0.0.5] - 2023-03-30

### Fixed
- Fix installation instructions

## [0.0.4] - 2023-03-29

### Fixed
- Fix missing style file

## [0.0.3] - 2023-03-28

### Added
- Added `halodata` package to iterate over halo doppler lidar data from Cloudnet
- Added `haloboard` package for data visualisations
- Added tests for CF Convention compliance

[0.0.3]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.0.3
[0.0.4]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.0.4
[0.0.5]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.0.5
[0.0.6]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.0.6
[0.0.7]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.0.7
[0.0.8]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.0.8
[0.0.9]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.0.9
[0.1.0]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.1.0
[0.1.1]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.1.1
[0.1.2]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.1.2
[0.1.3]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.1.3
[0.1.4]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.1.4
[0.1.5]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.1.5
[0.1.6]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.1.6
[0.1.7]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.1.7
[0.1.8]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.1.8
[0.1.9]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.1.9
[0.1.9]: https://github.com/actris-cloudnet/halo-reader/releases/tag/v0.1.9


src/haloreader/cli.py
import argparse
import datetime
import logging
from glob import glob
from pathlib import Path

import matplotlib.pyplot as plt

from haloboard.writer import Writer
from halodata.datasets import get_halo_cloudnet
from haloreader.read import read, read_bg
from haloreader.type_guards import is_ndarray

log = logging.getLogger(__name__)


def halo_reader() -> None:
    logging.basicConfig(level=logging.INFO)
    args = _haloreader_args()
    if args.subcommand == "from_cloudnet":
        _from_cloudnet(args)
    elif args.subcommand == "from_raw":
        _from_raw(args)
    else:
        raise NotImplementedError


def _from_cloudnet(args: argparse.Namespace) -> None:
    halo, halobg = get_halo_cloudnet(site=args.site, date=args.date)
    if halo is None:
        log.warning("No data from %s on %s", args.site, args.date)
        return
    if halobg is None:
        raise TypeError
    log.info("Correct background")
    halo.correct_background(halobg)
    log.info("Compute beta")
    halo.compute_beta()
    log.info("Compute noise screen")
    screen = halo.compute_noise_screen()
    log.info("Compute screened beta")
    halo.compute_beta_screened(screen)
    log.info("Compute screened doppler velocity")
    halo.compute_doppler_velocity_screened(screen)
    log.info("Convert timeunits")
    halo.convert_time_unit2cloudnet_time()
    log.info("Create netCDF")
    nc_buff = halo.to_nc()
    with open(f"halo_{args.site}_{args.date}.nc", "wb") as f:
        f.write(nc_buff)
    if args.plot:
        log.info("Create plots")
        writer = Writer()
        fig, ax = plt.subplots(nplots := 6, 1, figsize=(24, nplots * 6))
        halo.intensity_raw.plot(ax[0])
        halo.doppler_velocity.plot(ax[1])
        if halo.intensity is not None:
            halo.intensity.plot(ax[2])
        if halo.beta is not None:
            halo.beta.plot(ax[3])
        if halo.beta_screened is not None:
            halo.beta_screened.plot(ax[4])
        if halo.doppler_velocity_screened is not None:
            halo.doppler_velocity_screened.plot(ax[5])

        writer.add_figure(f"halo_{args.site}_{args.date}", fig)


def _parse_files_from_arg(path_list: list) -> list:
    allowed_wildcards = "*?"
    files = []
    for src in path_list:
        if any((c in allowed_wildcards) for c in src.name):
            wildcard_files = glob(src.name)
            if wildcard_files:
                for wildcard_file in wildcard_files:
                    files.append(Path(wildcard_file))
        else:
            files.append(src)
    return files


def _from_raw(args: argparse.Namespace) -> None:
    halo_src = [src for src in args.src if src.name.endswith(".hpl")]
    halo_src = _parse_files_from_arg(halo_src)
    bg_src = [
        src
        for src in args.src
        if src.name.startswith("Background") and src.name.endswith(".txt")
    ]
    bg_src = _parse_files_from_arg(bg_src)
    halo = read(halo_src)
    if halo is None:
        log.warning("No data")
        return

    halobg = read_bg(bg_src)
    if halobg:
        if not is_ndarray(halobg.background.data):
            raise TypeError
        if halobg.background.data.shape[0] < 300:
            log.warning(
                "Few (< 300) background profiles "
                "might lead to incorrect background correction"
            )
        halo.correct_background(halobg)
    else:
        log.warning("No background files, skipping background correction")
    nc_buff = halo.to_nc()
    with args.output.open("wb") as f:
        f.write(nc_buff)
    if args.plot:
        writer = Writer()
        fig, ax = plt.subplots(3, 1, figsize=(24, 16))
        halo.intensity_raw.plot(ax[0])
        halo.doppler_velocity.plot(ax[1])
        if halo.intensity:
            halo.intensity.plot(ax[2])
        writer.add_figure(f"{args.output.stem}", fig)


def _haloreader_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    subparsers = parser.add_subparsers(
        title="subcommands", dest="subcommand", required=True
    )
    _from_cloudnet_args(
        subparsers.add_parser(
            "from_cloudnet", help="Download files directly from cloudnet."
        )
    )
    _from_raw_args(
        subparsers.add_parser(
            "from_raw", help="Read and background correct raw files into netCDF"
        )
    )

    return parser.parse_args()


def _from_cloudnet_args(parser: argparse.ArgumentParser) -> None:
    parser.add_argument("-p", "--plot", action="store_true")
    parser.add_argument("-s", "--site", type=str, default="warsaw")
    parser.add_argument(
        "-d",
        "--date",
        type=datetime.date.fromisoformat,
        default=datetime.date.today() - datetime.timedelta(days=1),
    )


def _from_raw_args(parser: argparse.ArgumentParser) -> None:
    parser.add_argument("-p", "--plot", action="store_true")
    parser.add_argument(
        "src",
        type=Path,
        nargs="*",
        default=[],
        help="Raw and background files in an arbitrary order",
    )
    parser.add_argument("-o", "--output", type=Path, required=True)
